### ISSUE TITLE
[BUG] writing parquet with partitionBy() loses sort order

### ISSUE THREAD

**Issue by etseidl** [NVIDIA/spark-rapids#3002]
**Describe the bug**
I'm trying to read from partitioned parquet, sort within partitions, and then write each partition back out.  Everything works great up until the last step where I call partitionBy().  The resulting file is no longer fully sorted.  Omitting the partitionBy yields the correct result, but without the partitioning.

**Steps/Code to reproduce bug**
Run the following from a spark shell

```scala
val df = sc.parallelize((1L to 10000000L)).map{i => ("a", f"$i%010d", i)}.toDF("partkey", "val", "val2")
df.repartition(1, $"partkey").sortWithinPartitions($"partkey", $"val", $"val2").write.mode("overwrite").partitionBy("partkey").parquet("file:///tmp/foosort")
df.repartition(1, $"partkey").sortWithinPartitions($"partkey", $"val", $"val2").write.mode("overwrite").parquet("file:///tmp/foosort2")
```

**Expected behavior**
After running the above, foosort will be partially sorted, but not completely.  foosort2 is fully sorted as expected.

**Environment details (please complete the following information)**
This happens both standalone and running via YARN.  I'm using branch 21.08 from around 6/18/2021, with spark 3.2.0-SNAPSHOT.  I set spark.sql.adaptive.enabled=false
 

**Additional context**
The problem seems to go away if I change the "val" column to an Int.  I also notice that the plans in the spark GUI differ.  Without the partitionBy the last stage does GpuColumnarExchange->GpuShuffleCoalesce->GpuSort.  With partitionBy there are extra steps GpuColumnarExchange->GpuShuffleCoalesce->GpuSort->GpuCoalesceBatches->GpuProject->GpuSort.

**Comment by jlowe** [comment]
Thanks for the bug report and excellent repro case, @etseidl!  I can reproduce this locally on 21.08-SNAPSHOT.

**Comment by jlowe** [comment]
I tracked this down to a failure to propagate the child sort order in `GpuCoalesceBatches`.  Essentially this bug caused the plan to "forget" that the data was already sorted on partkey, val, and val2, and it thought it needed to sort it again, this time just on the partkey.  Unless stable sorting is enabled in the conf, the sort is not guaranteed to be stable and sometimes the other values could appear out-of-order after the partkey-only sort.

By properly propagating the child sort order in the plan, the partitioned write realizes it doesn't need to sort again and the data comes out in the expected order.

> The problem seems to go away if I change the "val" column to an Int.

I suspect this is not always guaranteed since the sort is not stable.

> With partitionBy there are extra steps GpuColumnarExchange->GpuShuffleCoalesce->GpuSort->GpuCoalesceBatches->GpuProject->GpuSort.

This extra sort is caused by the lack of output ordering propagation, forcing the partitioned write to sort by partition key so it can batch the write accordingly.  This is fixed by #3009.

**Comment by etseidl** [comment]
tested the fix with my data and it worked!  thanks @jlowe
